pkg/codegen/internal/test/helpers.go

```diff
@@ -76,55 +78,85 @@ func LoadFiles(dir, lang string, files []string) (map[string][]byte, error) {
 	return result, nil
 }
 
-func loadDirectory(fs map[string][]byte, root, path string) error {
-	entries, err := os.ReadDir(path)
+func PathExists(path string) (bool, error) {
+	_, err := os.Stat(path)
+
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+
+	if err == nil {
+		return true, nil
+	}
+
+	return false, err
+}
+
+// `LoadBaseline` loads the contents of the given baseline directory,
+// by inspecting its `codegen-manifest.json`.
+func LoadBaseline(dir, lang string) (map[string][]byte, error) {
+	cm := &codegenManifest{}
+	err := cm.load(filepath.Join(dir, lang))
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("Failed to load codegen-manifest.json: %w", err)
 	}
 
-	for _, e := range entries {
-		entryPath := filepath.Join(path, e.Name())
-		if e.IsDir() {
-			if err = loadDirectory(fs, root, entryPath); err != nil {
-				return err
-			}
-		} else {
-			contents, err := os.ReadFile(entryPath)
-			if err != nil {
-				return err
-			}
-			name := filepath.ToSlash(entryPath[len(root)+1:])
+	files := make(map[string][]byte)
 
-			fs[name] = contents
+	for _, f := range cm.EmittedFiles {
+		bytes, err := ioutil.ReadFile(filepath.Join(dir, lang, f))
+		if err != nil {
+			return nil, fmt.Errorf("Failed to load file %s referenced in codegen-manifest.json: %w", f, err)
 		}
+		files[f] = bytes
 	}
 
-	return nil
+	return files, nil
 }
 
-// LoadBaseline loads the contents of the given baseline directory.
-func LoadBaseline(dir, lang string) (map[string][]byte, error) {
-	dir = filepath.Join(dir, lang)
+type codegenManifest struct {
+	EmittedFiles []string `json:"emittedFiles"`
+}
 
-	fs := map[string][]byte{}
-	if err := loadDirectory(fs, dir, dir); err != nil {
-		return nil, err
+func (cm *codegenManifest) load(dir string) error {
+	bytes, err := ioutil.ReadFile(filepath.Join(dir, "codegen-manifest.json"))
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(bytes, cm)
+}
+
+func (cm *codegenManifest) save(dir string) error {
+	sort.Strings(cm.EmittedFiles)
+	buf := &bytes.Buffer{}
+	enc := json.NewEncoder(buf)
+	enc.SetIndent("", "  ")
+	err := enc.Encode(cm)
+	if err != nil {
+		return err
 	}
-	return fs, nil
+	data := buf.Bytes()
+	return ioutil.WriteFile(filepath.Join(dir, "codegen-manifest.json"), data, 0600)
```

Comment:
Nit: rather than emitting to an in-memory buffer, we could emit directly to the file:

```go
	sort.Strings(cm.EmittedFiles)

	f, err := os.Create(filepath.Join(dir, "codegen-manifest.json"), 0600)
	if err != nil {
		return err
	}
	defer f.Close()

 	enc := json.NewEncoder(f)
 	enc.SetIndent("", "  ")
	return enc.Encode(cm)
```